### PR TITLE
FHT: Add email verification step for business trial plan

### PR DESCRIPTION
**Pull request by vykes-mac**
Closes https://github.com/Automattic/dotcom-forge/issues/4058

## Proposed Changes
This PR adds an additional step to verify emails when a user email is unverified before they can proceed with the hosting free trial.

This is similar to what's done in the migration free trial, I wanted to re-use that code but opted to create own step instead of trying to refactor migration trial step to match both needs. We might add additional things or change it so I think it's best to create different files.

Also the styling might need some work and it seems from p1696387524534499-slack-C041RHH38NQ that it is not finalise as yet, however I think we can start the discussion with this PR

* Add email verification step
* Show this step only when the business free trial plan is selected

## Testing Instructions

* Have a user with unverified email
* Go to `/setup/new-hosted-site` and select the business free trial plan
* Verify the email verification step is shown
* Resend the email then verify in your email address, 
* Verify email verification step auto refreshes and continues to checkout.

image

## Pre-merge Checklist

- [ ] Has the general commit checklist been followed? (PCYsg-hS-p2)
- [ ] https://wpcalypso.wordpress.com/devdocs/docs/testing/index.md for your changes?
- [ ] Have you tested the feature in Simple (P9HQHe-k8-p2), Atomic (P9HQHe-jW-p2), and self-hosted Jetpack sites (PCYsg-g6b-p2)?
- [ ] Have you checked for TypeScript, React or other console errors?
- [ ] Have you used memoizing on expensive computations? More info in [Memoizing with create-selector](https://github.com/Automattic/wp-calypso/blob/trunk/packages/state-utils/src/create-selector/README.md) and [Using memoizing selectors](https://react-redux.js.org/api/hooks#using-memoizing-selectors) and [Our Approach to Data](https://github.com/Automattic/wp-calypso/blob/trunk/docs/our-approach-to-data.md)
- [ ] Have we added the "[Status] String Freeze" label as soon as any new strings were ready for translation (p4TIVU-5Jq-p2)?
- [ ] For changes affecting Jetpack: Have we added the "[Status] Needs Privacy Updates" label if this pull request changes what data or activity we track or use (p4TIVU-ajp-p2)?

### CODE REVIEW

File: client/landing/stepper/declarative-flow/new-hosted-site-flow.ts [hunk 1]
```diff
@@ -89,6 +96,14 @@ const hosting: Flow = {
 						extra: { geo_affinity: siteGeoAffinity },
 					} );
 
+					if ( productSlug === 'wp_bundle_hosting_trial_monthly' ) {
```
Comment (zaguiini) [inline client/landing/stepper/declarative-flow/new-hosted-site-flow.ts]: Let's use the constant: 

```suggestion
					if ( productSlug === PLAN_HOSTING_TRIAL_MONTHLY ) {
```

File: client/landing/stepper/declarative-flow/internals/steps-repository/hosting-verify-email/index.tsx [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+import { EVERY_FIVE_SECONDS, Interval } from 'calypso/lib/interval';
+import { useSelector } from 'calypso/state';
+import { fetchCurrentUser } from 'calypso/state/current-user/actions';
+import { getCurrentUser } from 'calypso/state/current-user/selectors';
+import VerifyEmail from './verify-email';
+import type { Step } from '../../types';
+import type { UserData } from 'calypso/lib/user/user';
+import './style.scss';
+
+const HostingVerifyEmail: Step = function ImportVerifyEmail( { navigation } ) {
```
Comment (zaguiini) [inline client/landing/stepper/declarative-flow/internals/steps-repository/hosting-verify-email/index.tsx]: As per the linked Slack discussion, I think we should move away from calling it "e-mail verification step" and call it "trial description and e-mail verification step". Obviously not named like that 😅 Maybe "HostingTrialDescription" or "Conditions" would be better?

**Review by zaguiini — COMMENTED** [review]
What if we design the step this way:

"Start the Business trial

After upgrading to a paid plan, the following...
- X
- Y
- Z

[Try the Business plan for 3 days]"

That's the scenario where the user email is confirmed. If it's not, we could disable the button and display a helper text saying that the email needs to be verified, along with a link/button that resends the email if necessary.

**Comment by vykes-mac** [comment]
I feel like that's adding more friction by adding another screen just to say "Start business trial" after selecting a plan that basically is saying start trial. 

I understand we want to list the limitations of the trial but why not have that displayed in the plan selection itself and only show this e-mail verification if they haven't verified their email. 

It's one less step to me.

**Comment by zaguiini** [comment]
>I understand we want to list the limitations of the trial but why not have that displayed in the plan selection itself

In https://github.com/Automattic/wp-calypso/pull/82690, we're going to display the "try for free" CTA inside the Business-plan card. Unless we find a way to communicate the limitations in a way that is clear to the user, I think we'll need to always display the step.

>I feel like that's adding more friction by adding another screen just to say "Start business trial" after selecting a plan that basically is saying start trial. [...] It's one less step to me.

I understand that, but is it a big deal? They're getting it for free, and will just ack the limitations. We've been able to squeeze a site options step that is basically useless and still has 90% conversion, so 🤷 What I'm trying to say is that it's not a big deal, but I might be convinced otherwise.

**Comment by vykes-mac** [comment]
I think the limitations can be added to the plan card .
image

**Comment by vykes-mac** [comment]
> In https://github.com/Automattic/wp-calypso/pull/82690, we're going to display the "try for free" CTA inside the Business-plan card. Unless we find a way to communicate the limitations in a way that is clear to the user, I think we'll need to always display the step.

Okay so we changing the plan grid. Cool, I thought it was weird that the plan says "Start trial" then we have another screen saying start trial

**Comment by zaguiini** [comment]
This is how it looks right now:

https://github.com/Automattic/wp-calypso/assets/26530524/ee19683c-c84b-4b9a-ad79-5c4ef1e8be70

The only thing that's missing is the e-mail verification logic (polling + resend email + enable the start trial button once the email is verified) inside the [\<CallToAction /> component](https://github.com/Automattic/wp-calypso/blob/add/email-verify-step/client/landing/stepper/declarative-flow/internals/steps-repository/trial-acknowledge/hosting-trial-acknowledge.tsx#L25).

We could take these existing components as inspiration, and even refactor them
- client/components/email-verification/email-verification-gate/email-unverified-notice.jsx
- client/my-sites/checkout/checkout-thank-you/atomic-store-thank-you-card.jsx
- client/landing/stepper/declarative-flow/internals/steps-repository/import-verify-email/index.tsx

Apart from that, it's features and limitations double-checking, and copy writing. But that can be done later.

**Comment by vykes-mac** [comment]
I've added email verification notice with a link button to resend email.
This message should automatically disappear after it's verified and the `start trial `button enabled
image

**Comment by mpkelly** [comment]
Looks good to me. It showed me the notice even though I used an account with a confirmed email. The notice would clear after five seconds. I fixed that issue and also translated the notice text. 

- ✅ resending works
- ✅ confirming email works
- ✅ only shown for new accounts
- ✅ existing accounts can continue as soon as the `user` is fetched

**Comment by mpkelly** [comment]
~~I tested this and it looks good. I created a site with a trail plan using the new button. I checked `/start/plans`, and there is no change as expected; same when going through the `setup/import-hosted-site~~

I added this comment to the wrong PR yesterday. Moved it to here https://github.com/Automattic/wp-calypso/pull/82690#issuecomment-1754259302

**Comment by vykes-mac** [comment]
> Looks good to me. It showed me the notice even though I used an account with a confirmed email. The notice would clear after five seconds. I fixed that issue and also translated the notice text.

Thanks for fixing that, I forgot to set the `isVerified` to `user.email_verified` instead of `false`

**Review by zaguiini — APPROVED** [review]
Let's go 👍 🚀

File: client/landing/stepper/declarative-flow/internals/steps-repository/trial-acknowledge/use-verify-email.tsx [hunk 3]
```diff
@@ -44,9 +44,9 @@ export const useVerifyEmail = () => {
 	};
 
 	useEffect( () => {
-		const interval = setInterval( () => {
+		const interval = setInterval( async () => {
```
Comment (mpkelly) [inline client/landing/stepper/declarative-flow/internals/steps-repository/trial-acknowledge/use-verify-email.tsx]: Just removed this `async`/`await` which wasn't needed in the final fix.

**Comment by a8ci18n** [comment]
This Pull Request is now available for translation here: https://translate.wordpress.com/deliverables/9304210

Thank you @vykes-mac for including a screenshot in the description! This is really helpful for our translators.

**Comment by a8ci18n** [comment]
Translation for this Pull Request [has now been finished](https://translate.wordpress.com/deliverables/overview/9304210).

**Pull request merged**